### PR TITLE
Week05

### PR DESCRIPTION
**Pull request by safiweb**
# Adding unit tests for specified set of code

This PR adds unit tests for the set of code in clause.go, store.go and errors.go files. The tests are table-driven tests using subtests.  

For clause.go, the unit tests are in clause_test.go. Here we have two unit test functions:

_**TestClauses_String**_ which calls the _Strings()_ method and run the sub test cases, comparing what we got with what we wanted.

```
E:\Projects\Go\src\gopherguides-intro-to-go\week05>go test -v -run TestClauses_String ./...
=== RUN   TestClauses_String
=== PAUSE TestClauses_String
=== CONT  TestClauses_String
=== RUN   TestClauses_String/no_clauses_provided
=== RUN   TestClauses_String/clauses_provided
=== RUN   TestClauses_String/clauses_provided_failure
    clause_test.go:26: unexpected value, got: "animals" = "Lion" and "sports" = "F1", exp: "animals" = "Lion" and "sports" = "F1"
--- FAIL: TestClauses_String (0.00s)
    --- PASS: TestClauses_String/no_clauses_provided (0.00s)
    --- PASS: TestClauses_String/clauses_provided (0.00s)
    --- FAIL: TestClauses_String/clauses_provided_failure (0.00s)
FAIL
FAIL    week05  0.340s
FAIL
```

**_TestClauses_Match_** which calls the _Match()_ method and run the sub test cases, comparing what we got with what we wanted.

```
E:\Projects\Go\src\gopherguides-intro-to-go\week05>go test -v -run TestClauses_Match ./...
=== RUN   TestClauses_Match
=== PAUSE TestClauses_Match
=== CONT  TestClauses_Match
=== RUN   TestClauses_Match/nil_data_provided
=== RUN   TestClauses_Match/matching_data_provided
=== RUN   TestClauses_Match/no_matching_data_provided
--- PASS: TestClauses_Match (0.00s)
    --- PASS: TestClauses_Match/nil_data_provided (0.00s)
    --- PASS: TestClauses_Match/matching_data_provided (0.00s)
    --- PASS: TestClauses_Match/no_matching_data_provided (0.00s)
PASS
ok      week05  0.194s
```

For store.go, the unit tests are in store_test.go. Here we have these four unit test functions:

**_TestStore_All_** which calls the _All_ method and run the sub test cases, comparing what we got with what we wanted. It asserts error and the _Models_ returned via _assertError_ and _assertEqualModels_ helper functions.

```
E:\Projects\Go\src\gopherguides-intro-to-go\week05>go test -v -run TestStore_All ./...
=== RUN   TestStore_All
=== PAUSE TestStore_All
=== CONT  TestStore_All
=== RUN   TestStore_All/wrong_table
    store_test.go:44: unexpected error: table not found people
=== RUN   TestStore_All/correct_table
=== RUN   TestStore_All/correct_table_wrong_data
    store_test.go:45: unexpected error: got [map[age:30 name:Tom]], want [map[age:30 name:Jesse]]
--- FAIL: TestStore_All (0.00s)
    --- FAIL: TestStore_All/wrong_table (0.00s)
    --- PASS: TestStore_All/correct_table (0.00s)
    --- FAIL: TestStore_All/correct_table_wrong_data (0.00s)
FAIL
FAIL    week05  0.217s
FAIL
```

**_TestStore_Len_** which calls the _Len_ method and run the sub test cases, comparing what we got with what we wanted. It asserts the error and the wanted value. The error is asserted via via _assertError_ helper function.

```
E:\Projects\Go\src\gopherguides-intro-to-go\week05>go test -v -run TestStore_Len ./...
=== RUN   TestStore_Len
=== PAUSE TestStore_Len
=== CONT  TestStore_Len
=== RUN   TestStore_Len/correct_table
=== RUN   TestStore_Len/wrong_table
    store_test.go:86: unexpected error: table not found users
=== RUN   TestStore_Len/correct_table_wrong_return
    store_test.go:88: unexpected error: got 2, want 1
--- FAIL: TestStore_Len (0.00s)
    --- PASS: TestStore_Len/correct_table (0.00s)
    --- FAIL: TestStore_Len/wrong_table (0.00s)
    --- FAIL: TestStore_Len/correct_table_wrong_return (0.00s)
FAIL
FAIL    week05  0.142s
FAIL
```

**_TestStore_Insert_** which calls the _Insert_ method and run the sub test cases, checking if the data was inserted into the table.

```
E:\Projects\Go\src\gopherguides-intro-to-go\week05>go test -v -run TestStore_Insert ./...
=== RUN   TestStore_Insert
=== PAUSE TestStore_Insert
=== CONT  TestStore_Insert
=== RUN   TestStore_Insert/empty_store_insertion
=== RUN   TestStore_Insert/existing_store_insertion
--- PASS: TestStore_Insert (0.00s)
    --- PASS: TestStore_Insert/empty_store_insertion (0.00s)
    --- PASS: TestStore_Insert/existing_store_insertion (0.00s)
PASS
ok      week05  0.215s
```

**_TestStore_Select_** which calls the _Select_ method and run the sub test cases, comparing what we got with what we wanted. Checking if the provided query clauses returned correct data and table. It asserts error and the _Models_ returned via _assertError_ and _assertEqualModels_ helper functions.

```
E:\Projects\Go\src\gopherguides-intro-to-go\week05>go test -v -run TestStore_Select ./...
=== RUN   TestStore_Select
=== PAUSE TestStore_Select
=== CONT  TestStore_Select
=== RUN   TestStore_Select/empty_clauses_on_empty_store
    store_test.go:238: unexpected error: table not found users
=== RUN   TestStore_Select/empty_clauses_on_existing_store
    store_test.go:239: unexpected error: got [map[age:30 name:Tom]], want []
=== RUN   TestStore_Select/wrong_row_selection
    store_test.go:238: unexpected error: [users] no rows found
        query: "animals" = "Lion" and "me" = "Anthony" and "sports" = "F1"
=== RUN   TestStore_Select/correct_row_selection
=== RUN   TestStore_Select/wrong_return_-_correct_row_selection
    store_test.go:239: unexpected error: got [map[age:30 name:Tom]], want [map[age:40 name:Tom]]
--- FAIL: TestStore_Select (0.00s)
    --- FAIL: TestStore_Select/empty_clauses_on_empty_store (0.00s)
    --- FAIL: TestStore_Select/empty_clauses_on_existing_store (0.00s)
    --- FAIL: TestStore_Select/wrong_row_selection (0.00s)
    --- PASS: TestStore_Select/correct_row_selection (0.00s)
    --- FAIL: TestStore_Select/wrong_return_-_correct_row_selection (0.00s)
FAIL
FAIL    week05  0.216s
FAIL
```

For errors.go, the unit tests are in errors_test.go. Here we have these six unit test functions:

**_TestErrTableNotFound_TableNotFound_**  which calls the _TableNotFound_ function and run the sub test cases, comparing what we got with what we wanted.

```
E:\Projects\Go\src\gopherguides-intro-to-go\week05>go test -v -run TestErrTableNotFound_TableNotFound ./...
=== RUN   TestErrTableNotFound_TableNotFound
=== PAUSE TestErrTableNotFound_TableNotFound
=== CONT  TestErrTableNotFound_TableNotFound
=== RUN   TestErrTableNotFound_TableNotFound/empty_ErrTableNotFound
=== RUN   TestErrTableNotFound_TableNotFound/ErrTableNotFound_with_correct_table
=== RUN   TestErrTableNotFound_TableNotFound/ErrTableNotFound_with_wrong_table
    errors_test.go:29: unexpected error: got people, want users
--- FAIL: TestErrTableNotFound_TableNotFound (0.00s)
    --- PASS: TestErrTableNotFound_TableNotFound/empty_ErrTableNotFound (0.00s)
    --- PASS: TestErrTableNotFound_TableNotFound/ErrTableNotFound_with_correct_table (0.00s)
    --- FAIL: TestErrTableNotFound_TableNotFound/ErrTableNotFound_with_wrong_table (0.00s)
FAIL
FAIL    week05  0.201s
FAIL
```

**_TestErrTableNotFound_IsErrTableNotFound_** which calls the _IsErrTableNotFound_ function and run the sub test cases, comparing what we got with what we wanted.

```
E:\Projects\Go\src\gopherguides-intro-to-go\week05>go test -v -run TestErrTableNotFound_IsErrTableNotFound ./...
=== RUN   TestErrTableNotFound_IsErrTableNotFound
=== PAUSE TestErrTableNotFound_IsErrTableNotFound
=== CONT  TestErrTableNotFound_IsErrTableNotFound
=== RUN   TestErrTableNotFound_IsErrTableNotFound/not_ErrTableNotFound_error
=== RUN   TestErrTableNotFound_IsErrTableNotFound/not_not_ErrTableNotFound_error
=== RUN   TestErrTableNotFound_IsErrTableNotFound/empty_ErrTableNotFound_error
--- PASS: TestErrTableNotFound_IsErrTableNotFound (0.00s)
    --- PASS: TestErrTableNotFound_IsErrTableNotFound/not_ErrTableNotFound_error (0.00s)
    --- PASS: TestErrTableNotFound_IsErrTableNotFound/not_not_ErrTableNotFound_error (0.00s)
    --- PASS: TestErrTableNotFound_IsErrTableNotFound/empty_ErrTableNotFound_error (0.00s)
PASS
ok      week05  0.183s
```

**_TestErrNoRows_Clauses_** which calls the _Clauses_ method and run the sub test cases, comparing what we got with what we wanted.

```
E:\Projects\Go\src\gopherguides-intro-to-go\week05>go test -v -run TestErrNoRows_Clauses ./...
=== RUN   TestErrNoRows_Clauses
=== PAUSE TestErrNoRows_Clauses
=== CONT  TestErrNoRows_Clauses
=== RUN   TestErrNoRows_Clauses/empty_clauses_provided
    errors_test.go:83: unexpected error: got , want "hobby" = "golf" and "name" = "Webbs"
=== RUN   TestErrNoRows_Clauses/clauses_provided_correct_expected
=== RUN   TestErrNoRows_Clauses/clauses_provided_wrong_expected
    errors_test.go:83: unexpected error: got "continent" = "Africa" and "name" = "Lion", want "hobby" = "golf" and "name" = "Webbs"
--- FAIL: TestErrNoRows_Clauses (0.00s)
    --- FAIL: TestErrNoRows_Clauses/empty_clauses_provided (0.00s)
    --- PASS: TestErrNoRows_Clauses/clauses_provided_correct_expected (0.00s)
    --- FAIL: TestErrNoRows_Clauses/clauses_provided_wrong_expected (0.00s)
FAIL
FAIL    week05
```

**_TestErrNoRows_RowNotFound_** which calls the _RowNotFound_ method and run the sub test cases, comparing what we got with what we wanted.

```
E:\Projects\Go\src\gopherguides-intro-to-go\week05>go test -v -run TestErrNoRows_RowNotFound ./...
=== RUN   TestErrNoRows_RowNotFound
=== PAUSE TestErrNoRows_RowNotFound
=== CONT  TestErrNoRows_RowNotFound
=== RUN   TestErrNoRows_RowNotFound/empty_clauses_and_table_name
=== RUN   TestErrNoRows_RowNotFound/empty_clauses,_correct_table_name_provided
=== RUN   TestErrNoRows_RowNotFound/clauses_provided,_wrong_table_wanted
    errors_test.go:111: unexpected error: got users, want people
=== RUN   TestErrNoRows_RowNotFound/correct_clauses,_correct_table_provided
--- FAIL: TestErrNoRows_RowNotFound (0.00s)
    --- PASS: TestErrNoRows_RowNotFound/empty_clauses_and_table_name (0.00s)
    --- PASS: TestErrNoRows_RowNotFound/empty_clauses,_correct_table_name_provided (0.00s)
    --- FAIL: TestErrNoRows_RowNotFound/clauses_provided,_wrong_table_wanted (0.00s)
    --- PASS: TestErrNoRows_RowNotFound/correct_clauses,_correct_table_provided (0.00s)
FAIL
FAIL    week05  0.188s
FAIL
```

**_TestErrNoRows_IsErrNoRows_** which calls the _IsErrNoRows_ function and run the sub test cases, comparing what we got with what we wanted. 

```
E:\Projects\Go\src\gopherguides-intro-to-go\week05>go test -v -run TestErrNoRows_IsErrNoRows ./...
=== RUN   TestErrNoRows_IsErrNoRows
=== PAUSE TestErrNoRows_IsErrNoRows
=== CONT  TestErrNoRows_IsErrNoRows
=== RUN   TestErrNoRows_IsErrNoRows/not_errNoRows_error_type
    errors_test.go:143: unexpected error: got false, want true
=== RUN   TestErrNoRows_IsErrNoRows/errNoRows_error_type,_empty_clauses
=== RUN   TestErrNoRows_IsErrNoRows/errNoRows_error_type,_correct_data
--- FAIL: TestErrNoRows_IsErrNoRows (0.00s)
    --- FAIL: TestErrNoRows_IsErrNoRows/not_errNoRows_error_type (0.00s)
    --- PASS: TestErrNoRows_IsErrNoRows/errNoRows_error_type,_empty_clauses (0.00s)
    --- PASS: TestErrNoRows_IsErrNoRows/errNoRows_error_type,_correct_data (0.00s)
FAIL
FAIL    week05  0.612s
FAIL
```

**_TestErrNoRows_AsErrNoRows_** which calls the _AsErrNoRows_ method and run the sub test cases, comparing what we got with what we wanted.

```
E:\Projects\Go\src\gopherguides-intro-to-go\week05>go test -v -run TestErrNoRows_AsErrNoRows ./...
=== RUN   TestErrNoRows_AsErrNoRows
=== PAUSE TestErrNoRows_AsErrNoRows
=== CONT  TestErrNoRows_AsErrNoRows
=== RUN   TestErrNoRows_AsErrNoRows/as_errNoRows_no_
    errors_test.go:172: unexpected error: got <nil>, want [] no rows found
        query:
=== RUN   TestErrNoRows_AsErrNoRows/as_errNoRows_yes,_empty_clauses
=== RUN   TestErrNoRows_AsErrNoRows/as_errNoRows_yes,_wrongclauses_provided
--- FAIL: TestErrNoRows_AsErrNoRows (0.00s)
    --- FAIL: TestErrNoRows_AsErrNoRows/as_errNoRows_no_ (0.00s)
    --- PASS: TestErrNoRows_AsErrNoRows/as_errNoRows_yes,_empty_clauses (0.00s)
    --- PASS: TestErrNoRows_AsErrNoRows/as_errNoRows_yes,_wrongclauses_provided (0.00s)
FAIL
FAIL    week05  0.168s
FAIL
```

The tests achieved 100% coverage as seen below:

image

## Changes Summary
Added clause.go, store.go, errors.go that have specified assignment code
Added clause_test.go, store_test.go and errors_test.go that have unit test functions
Miscellaneous - formatting code

## Difficulties
- Calling  t.Parallel() in sub tests brought different test results and also made test coverage not achieavable. It's similar as explained here [https://gist.github.com/posener/92a55c4cd441fc5e5e85f27bca008721](https://gist.github.com/posener/92a55c4cd441fc5e5e85f27bca008721). I had to revert to the conventional way i.e calling t.Parallel() at the very top of the test function.

- A method or function returns more than one result - testing for only one result made the test coverage to be minimal. Had to test for all results.

## Surprises

Table-driven tests using subtests are more cleaner and enjoyable to write.

### CODE REVIEW

**Review by markbates — CHANGES_REQUESTED** [review]
Please update the `.github/workflows/tests.yml` file in this branch so the tests will run.

File: week05/go.mod [hunk 1]
```diff
@@ -0,0 +1,3 @@
+module week05
```
Comment (markbates) [inline week05/go.mod]: Please remove the `week05/go.mod` file so your tests will run.

File: week05/errors_test.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+		{"clauses provided correct expected", &errNoRows{Clauses{"hobby": "golf", "name": "Webbs"}, "users"}, Clauses{"name": "Webbs", "hobby": "golf"}},
+		{"clauses provided wrong expected", &errNoRows{Clauses{"name": "Lion", "continent": "Africa"}, "animals"}, Clauses{"hobby": "golf", "name": "Webbs"}},
+	}
+
+	for _, tt := range testcases {
+		t.Run(tt.name, func(t *testing.T) {
+
+			got := tt.e.Clauses()
+
+			if !reflect.DeepEqual(got, tt.want) {
```
Comment (markbates) [inline week05/errors_test.go]: Please remove all usages of `reflect`. You do not need it to complete this, or any, assignment.

File: week05/errors_test.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+	t.Parallel()
+
+	testcases := []struct {
+		name string
+		e    *errNoRows
+		want Clauses
+	}{
+		{"empty clauses provided", &errNoRows{Clauses{}, "users"}, Clauses{"hobby": "golf", "name": "Webbs"}},
+		{"clauses provided correct expected", &errNoRows{Clauses{"hobby": "golf", "name": "Webbs"}, "users"}, Clauses{"name": "Webbs", "hobby": "golf"}},
+		{"clauses provided wrong expected", &errNoRows{Clauses{"name": "Lion", "continent": "Africa"}, "animals"}, Clauses{"hobby": "golf", "name": "Webbs"}},
```
Comment (markbates) [inline week05/errors_test.go]: Always use field names when initializing a struct. If the order of the fields changes, or a new one is added, your code will fail to compile. 

Please update all instances that do not use struct fields.

**Review by markbates — CHANGES_REQUESTED** [review]
Good work overall.

Please fix the issues I've outlined and re-submit.

File: week05/go.mod (hunk 1 above)
Comment (safiweb) [reply week05/go.mod]: ok

File: week05/errors_test.go (hunk 3 above)
Comment (safiweb) [reply week05/errors_test.go]: ok

File: week05/errors_test.go (hunk 2 above)
Comment (safiweb) [reply week05/errors_test.go]: ok

File: week05/go.mod (hunk 1 above)
Comment (safiweb) [reply week05/go.mod]: @markbates I have removed the week05/go.mod and updated tests.yml but now the PR tests checks are failing. Please advise.

**Review by markbates — COMMENTED** [review]
Anthony, I'm confused by this PR. There seems to be work done across multiple weeks. 

This PR should only focus on this week's assignment.

Please clean up this PR.

**Pull request closed without merging**